exp/image/colour.d

```diff
@@ -0,0 +1,695 @@
+				ubyte b = (val(hex[4]) << 4) | val(hex[5]);
+				this = Colour!("rgb", ubyte, colourSpace)(r, g, b);
+			}
+			if(hex.length == 8)
+			{
+				ubyte a = (val(hex[0]) << 4) | val(hex[1]);
+				ubyte r = (val(hex[2]) << 4) | val(hex[3]);
+				ubyte g = (val(hex[4]) << 4) | val(hex[5]);
+				ubyte b = (val(hex[6]) << 4) | val(hex[7]);
+				this = Colour!("rgb", ubyte, colourSpace)(r, g, b, a);
+			}
+			else
+				assert(false, "Invalid hex string!");
+		}
+	}
+
+	void opAssign(C)(C v) if(!is(C == typeof(this)) && isColour!C)
+	{
+		this = v.to!(typeof(this));
+	}
+
+	C opCast(C)() if(isColour!C)
+	{
+		return to!C(this);
+	}
+
+	static if(hasAlpha)
+		auto tristimulusWithAlpha() { return tuple(tristimulus.expand, a); }
+	else
+		auto tristimulusWithAlpha() { return tuple(tristimulus.expand, ComponentType(0)); }
+}
+
+
+// declare some common colour types...
+
+alias Colour!("rgb", ubyte) RGB;
+alias Colour!("rgbx", ubyte) RGBX;
+alias Colour!("rgba", ubyte) RGBA;
+
+alias Colour!("bgr", ubyte) BGR;
+alias Colour!("bgrx", ubyte) BGRX;
+alias Colour!("bgra", ubyte) BGRA;
+
+alias Colour!("rgb", ushort) RGB16;
+alias Colour!("rgbx", ushort) RGBX16;
+alias Colour!("rgba", ushort) RGBA16;
+
+alias Colour!("l", ubyte) L8;
+alias Colour!("l", ushort) L16;
+alias Colour!("la", ubyte) LA;
+alias Colour!("la", ushort) LA16;
+
+alias Colour!("rg", byte) UV8;
+
+alias Colour!("xyz", float, ColourSpace.XYZ) XYZ;
+
+
+// TODO: represent packed colours, eg R5G6B5, etc
+struct PackedColour(string format, ColourSpace colourSpace = ColourSpace.sRGB)
+{
+	// TODO: we'll try and fabricate the packing algorithm based on the format string
+	//...
+
+	void pack(C)(C colour);
+	C unpack(C)();
+}
+
+// std.conv style to! function, which performs almost all conversion operations
+To to(To, From)(From colour) if(isColour!To && isColour!From)
+{
+	static if(is(To == Colour!(cmpT, TyT, csT), string cmpT, TyT, ColourSpace csT) && is(From == Colour!(cmpF, TyF, csF), string cmpF, TyF, ColourSpace csF))
+	{
+		auto src = colour.tristimulusWithAlpha;
+
+		static if(csF == csT)
+		{
+			// colour space is the same, just do type conversion
+			return To(convert!TyT(src[0]), convert!TyT(src[1]), convert!TyT(src[2]), convert!TyT(src[3]));
+		}
+		else
+		{
+			static if(false && TyF.sizeof <= 2 && TyT.sizeof <= 2)
+			{
+				// <= 16bit type conversion should use a look-up table
+
+				// TODO: this is blocked by '^^' not working in CTFE!
+			}
+			else
+			{
+				// full colour space conversion
+
+				// unpack the working values
+				alias WorkType = WorkingType!(TyF, TyT);
+				WorkType a = convert!WorkType(src[0]);
+				WorkType b = convert!WorkType(src[1]);
+				WorkType c = convert!WorkType(src[2]);
+
+				// perform the colour space conversion
+				auto r = convertColourSpace!(csF, csT)(a, b, c);
+
+				// convert and return the output
+				return To(convert!TyT(a), convert!TyT(b), convert!TyT(c), convert!TyT(src[3]));
+			}
+		}
+	}
```

Comment:
It's impossible, or the constraint would fail.
The static if is only for the is() expressions, so I can give names to the inferred template arguments.
Is there a way I can do that without the static if?